mender.go

```diff
@@ -529,20 +533,48 @@ func shouldTransit(from, to State) bool {
 	return from.Transition() != to.Transition()
 }
 
-func (m *mender) TransitionState(to State, ctx *StateContext) (State, bool) {
-	from := m.GetCurrentState()
-	return m.transitionState(from, to, ctx)
-}
-
-func TransitionError(s State) State {
+func TransitionError(s State, action string) State {
 	me := NewTransientError(errors.New("error executing state script"))
-	if us, ok := s.(UpdateState); ok {
-		return NewUpdateErrorState(me, us.Update())
+	log.Errorf("will transit to error state form: %s [%s]",
+		s.Id().String(), s.Transition().String())
+	switch t := s.(type) {
+	case *UpdateFetchState:
+		new := NewUpdateStatusReportState(t.update, client.StatusFailure)
+		new.SetTransition(ToError)
+		return new
+	case *UpdateStoreState:
+		if action == "Leave" {
+			new := NewUpdateStatusReportState(t.update, client.StatusFailure)
+			new.SetTransition(ToError)
+			return new
+		}
+		return NewUpdateErrorState(me, t.update)
+	case *UpdateInstallState:
+		return NewRollbackState(t.Update(), false, false)
```

Comment:
While you're doing a change that big, you might want to refactor NewRollbackState to take a struct:
```
type RollbackActions struct {
    SwapPartitions bool
    Reboot bool
}
```
otherwise it's hard to tell what `true, true` or `false, false` stands for. Then instead of 
`NewRollbackState(t.Update(), true, true)` you'd have:

```
NewRollbackState(t.Update(), RollbackActions{
    SwapPartitions: true,
    Reboot: true,
})
```